### PR TITLE
prov/efa: adopt wide completion APIs in EFA provider

### PR DESCRIPTION
**Pull request by wenduwan**
Migrate efa provider to wide completion APIs. Deprecate ibv_poll_cq
usage in favor of ibv_{start,next,end}_poll verbs.

Signed-off-by: wenduwan <wenduwdan@amazon.com>

In this change we migrate to wide completion APIs to take advantage of additional query capabilities of the extended CQ verbs. Specifically, usage of ibv_poll_cq verb is deprecated in favor of ibv_{start,next,end}_poll verbs.

### CODE REVIEW

File: prov/efa/configure.m4 [hunk 1]
```diff
@@ -148,6 +148,7 @@ AC_DEFUN([FI_EFA_CONFIGURE],[
 						 [
 							#include <stdarg.h>
 							#include <stddef.h>
+							#include <stdint.h>
```
Comment (wenduwan) [inline prov/efa/configure.m4:151]: According to `cmocka.h`
```c
 * These headers or their equivalents MUST be included prior to including
 * this header file.
 * @code
 * #include <stdarg.h>
 * #include <stddef.h>
 * #include <stdint.h>
 * #include <setjmp.h>
 * @endcode
 *
 * This allows test applications to use custom definitions of C standard
 * library functions and types.
```

File: prov/efa/src/efa_cq.c [hunk 2]
```diff
@@ -331,11 +340,119 @@ int efa_cq_open(struct fid_domain *domain_fid, struct fi_cq_attr *attr,
...
+
+	const struct CMUnitTest efa_cq_unit_tests[] = {
+		cmocka_unit_test_setup_teardown(test_efa_cq_readerr_happy, setup_readerr, teardown_readerr),
+		cmocka_unit_test_setup_teardown(test_efa_cq_readerr_sad, setup_readerr, teardown_readerr),
+	};
+
+	ret = cmocka_run_group_tests_name("efa cq unit tests", efa_cq_unit_tests, NULL, NULL);
+
+	return ret;
+}
```
Comment (wenduwan) [inline prov/efa/src/efa_cq.c:456]:
```xml
<?xml version="1.0" encoding="UTF-8" ?>
<testsuites>
  <testsuite name="efa cq unit tests" time="0.000" tests="2" failures="0" errors="0" skipped="0" >
    <testcase name="test_efa_cq_readerr_happy" time="0.000" >
    </testcase>
    <testcase name="test_efa_cq_readerr_sad" time="0.000" >
    </testcase>
  </testsuite>
</testsuites>
```

File: prov/efa/src/efa_cq.c [hunk 3]
```diff
@@ -331,11 +340,119 @@ int efa_cq_open(struct fid_domain *domain_fid, struct fi_cq_attr *attr,
...
+	struct fi_cq_err_entry *entry;
+	struct efa_cq *efa_cq;
+};
+
+static enum ibv_wc_opcode _read_send_code(struct ibv_cq_ex *current)
+{
+	return IBV_WC_SEND;
+};
+
+static int setup_readerr(void **state)
```
Comment (wenduwan) [inline prov/efa/src/efa_cq.c:366]: Mock out the objects directly. This eliminates dependencies on other function/API, and more importantly does not require EFA device to run the tests.

File: prov/efa/src/efa_cq.c [hunk 4]
```diff
@@ -121,65 +139,57 @@ static void efa_cq_read_data_entry(struct efa_wc *wc, int i, void *buf)
...
 		}
-		cq->read_entry(&wc, i, buf);
+
+		cq->read_entry(&wc, num_cqe, buf);
+		num_cqe++;
+
+		err = ibv_next_poll(cq->ibv_cq_ex);
 	}
 
+	err = (err == ENOENT) ? -FI_EAGAIN : (err > 0 ? -err : err);
```
Comment (wzamazon) [inline prov/efa/src/efa_cq.c:185]: This is too complicated for 1 line, can you wrap it into an inline function?

File: prov/efa/src/efa_cq.c [hunk 5]
```diff
@@ -121,65 +139,57 @@ static void efa_cq_read_data_entry(struct efa_wc *wc, int i, void *buf)
...
+
+	if (should_end_poll)
```
Comment (wzamazon) [inline prov/efa/src/efa_cq.c:187]: seems we can just used `if (!err)` and do not really need `should_end_poll`.

File: prov/efa/src/efa_cq.c [hunk 6]
```diff
@@ -331,11 +340,119 @@ int efa_cq_open(struct fid_domain *domain_fid, struct fi_cq_attr *attr,
...
+	int ret;
+
+	cmocka_set_message_output(CM_OUTPUT_XML);
+
+	const struct CMUnitTest efa_cq_unit_tests[] = {
+		cmocka_unit_test_setup_teardown(test_efa_cq_readerr_happy, setup_readerr, teardown_readerr),
+		cmocka_unit_test_setup_teardown(test_efa_cq_readerr_sad, setup_readerr, teardown_readerr),
+	};
+
+	ret = cmocka_run_group_tests_name("efa cq unit tests", efa_cq_unit_tests, NULL, NULL);
```
Comment (wzamazon) [inline prov/efa/src/efa_cq.c:453]: why create a new group instead of add the test to existing group?

File: prov/efa/src/rxr/rxr_ep.c [hunk 7]
```diff
@@ -1868,7 +1865,17 @@ static inline void rdm_ep_poll_ibv_cq(struct rxr_ep *ep,
...
+
+		err = ibv_next_poll(efa_cq->ibv_cq_ex);
+	}
+
+	if (err && err != ENOENT) {
+		err = err > 0 ? err : -err;
+		efa_eq_write_error(&ep->util_ep, err, err);
 	}
+
+	if (should_end_poll)
```
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_ep.c:1877]: similarly, can use `if (!err)`.

File: prov/efa/src/efa_cq.c (hunk 4 above)
Comment (wenduwan) [reply prov/efa/src/efa_cq.c:185]: Sure. Will address.

File: prov/efa/src/efa_cq.c (hunk 6 above)
Comment (wenduwan) [reply prov/efa/src/efa_cq.c:453]: I don't see a strong connection between these tests, and they are testing distinct aspects from what I can tell.

The more important reason is that, I find the recommended approach not scalable as we add more tests, i.e.
```
1. Add unit tests to the bottom of the c source file you wish to test.
2. Add headers to efa_unit_tests.h for each new unit test.
3. Create or find an existing cmocka test group in efa_unit_tests.c
4. Add your new unit test to the cmocka test group.
```

For functions called in `efa_unit_tests.c` we will need to declare them in `efa_unit_tests.h`, which requires more code. This is also restrictive for developers, e.g. it's hard to reuse static functions in `efa_xxx.c` (declaring static functions in `efa_unit_tests.h` is not as straight forward because of dependencies).

From my perspective it's better to be frugal with what we put in `efa_unit_tests.x`, and hide detailed test fixtures in respective source files. Modifying tests is easier when the related code is localized in one or two places and shared with few other tests.

Comment (wzamazon) [reply prov/efa/src/efa_cq.c:472]: > I don't see a strong connection between these tests, and they are testing distinct aspects from what I can tell.

They are tests of EFA provider, therefore is in a group named "efa unit tests". It would be confusing to see two groups of tests, one is named "efa unit tests", the other is named "efa cq unit tests". 

> For functions called in efa_unit_tests.c we will need to declare them in efa_unit_tests.h, which requires more code. 

Yes, that is 1 line for each test, and one will need to declare a function so that it can be called from `efa_unit_tests.c`

IMO this PR added more code (and confusion) to avoid that declaration. The function `run_efa_cq_tests` (15 lines of code) can be avoided if we add the test directly to existing group.

Comment (wenduwan) [reply prov/efa/src/efa_cq.c:472]: The existing `efa unit tests` should be split into smaller groups. Even though they all 'test efa provider', groups are logically independent from each other. For example, the [cmocka_run_group_tests()](https://api.cmocka.org/group__cmocka__exec.html#ga7c62fd0acf2235ce98268c28ee262a57) function has this signature:
```c
int cmocka_run_group_tests ( 
                const struct CMUnitTest group_tests[],
		CMFixtureFunction  	group_setup,
		CMFixtureFunction  	group_teardown 
)
```
It implies that one criteria for test group is that, tests in the same group should require the same setup and teardown process. Tests in this PR do not align with the existing tests and has its own requirements (e.g. the new tests do not require EFA device).

More importantly keeping all test related code in `efa_unit_tests` is not a scalable approach and leads to unnecessary code - yes, centralized code is easier to reuse, but enforcing it means the developer must share details though that's not used elsewhere. Instead the developer should decide the right level of encapsulation, and what she wants to expose.

Comment (wzamazon) [reply prov/efa/src/efa_cq.c:472]: > The existing efa unit tests should be split into smaller groups.

There are only 2 test in `efa unit tests`. I think split them into smaller groups is overkill.

> More importantly keeping all test related code in efa_unit_tests is not a scalable approach and leads to unnecessary code.

I am not sure I agree. In my opinion the current PR has unnecessary code: the function `run_efa_cq_test` is not necessary.

File: prov/efa/src/efa_cq.c [hunk 8]
```diff
@@ -118,68 +136,79 @@ static void efa_cq_read_data_entry(struct efa_wc *wc, int i, void *buf)
...
 }
 
+/**
+ * @brief Adjust an error code from CQ poll API, e.g. `ibv_start_poll`, `ibv_end_poll`.
+ * The returned error code must be 0 (success) or negative (error).
+ * As a special case, if input error code is ENOENT (there was no item on CQ), we should return -FI_EAGAIN.
+ * @param[in] err	Return value from `ibv_start_poll` or `ibv_end_poll`
+ * @returns Adjusted error code
+ */
+static inline ssize_t efa_cq_poll_adjust_error(ssize_t err) {
```
Comment (wzamazon) [inline prov/efa/src/efa_cq.c:146]: `efa_cq_poll_adjust_error` -> `efa_cq_ibv_poll_error_to_fi_error`

**Review by wzamazon — COMMENTED** [review]
Actual implementation looks good, only has a minor comments. I would prefer it as a separate commit, and have unit test as a separate commit.

The unit test refactor commit should be merged into unit test. It does not make sense to have a follow up commit to modify a previous commit.

File: prov/efa/src/efa_cq.c (hunk 8 above)
Comment (wenduwan) [reply prov/efa/src/efa_cq.c:146]: I want to clarify this comment
> The unit test refactor commit should be merged into unit test

My understanding is to squash both commits into one. Business logic change should be accompanied by unit tests.

Do you mean different?

Comment (wenduwan) [reply prov/efa/src/efa_cq.c]: Sounds good. I squashed both commits.

File: prov/efa/src/efa_cq.c [hunk 9]
```diff
@@ -331,11 +359,210 @@ int efa_cq_open(struct fid_domain *domain_fid, struct fi_cq_attr *attr,
...
+};
+
+static int setup_readfrom(struct efa_cq_readfrom_state *state)
+{
+	int err;
+	static struct efa_cq *efa_cq;
+
+	efa_cq = calloc(1, sizeof(struct efa_cq));
+	if (!efa_cq)
+		return -1;
```
Comment (wzamazon) [inline prov/efa/src/efa_cq.c:393]: `return -FI_ENOMEM`.

File: prov/efa/src/efa_cq.c [hunk 10]
```diff
@@ -331,11 +359,210 @@ int efa_cq_open(struct fid_domain *domain_fid, struct fi_cq_attr *attr,
...
+
+	err = efa_cq_initialize(efa_cq);
+	if (err)
+		return -1;
```
Comment (wzamazon) [inline prov/efa/src/efa_cq.c:397]: `return err`.

File: prov/efa/test/efa_unit_test_utils.h [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+/**
+ * @brief Initialize a rxr_ep struct
+ * @param[in] rxr_ep A pointer to `struct rxr_ep`.
+ * @param[in] efa_cq A pointer to `struct efa_cq`. This is used to inject values into rxr_ep.
+ * @param[in] efa_ep A pointer to `struct efa_ep`. This is used to inject values into rxr_ep.
+ * @returns Error code. 0 if successful, non-zero if failed.
+ */
+static int rxr_ep_initialize(struct rxr_ep *rxr_ep, struct efa_cq *efa_cq, struct efa_ep *efa_ep)
```
Comment (wzamazon) [inline prov/efa/test/efa_unit_test_utils.h:72]: why not just call `fi_endpoint` to create an ep?

Like we did in `https://github.com/ofiwg/libfabric/blob/62521e7eef653367e3a6b88c99be7173c4e7befa/prov/efa/test/efa_unit_test_ep.c#L53`

File: prov/efa/test/efa_unit_test_utils.h [hunk 12]
```diff
@@ -0,0 +1,113 @@
+#include <cmocka.h>
+#include <infiniband/verbs.h>
+#include "efa.h"
+
+/**
+ * @brief Initialize a efa_cq struct
+ * @param[in] efa_cq A pointer to `struct efa_cq`.
+ * @returns Error code. 0 if successful, non-zero if failed.
+ */
+static int efa_cq_initialize(struct efa_cq *efa_cq)
```
Comment (wzamazon) [inline prov/efa/test/efa_unit_test_utils.h:10]: Can we use `fi_cq_open` to create a real cq?

File: prov/efa/src/efa_cq.c [hunk 13]
```diff
@@ -331,11 +359,210 @@ int efa_cq_open(struct fid_domain *domain_fid, struct fi_cq_attr *attr,
...
 }
 
+
+#if EFA_UNIT_TEST
+
+#include "efa_unit_tests.h"
+#include "efa_unit_test_utils.h"
+
+struct efa_cq_readfrom_state {
+	struct efa_cq *efa_cq;
```
Comment (wzamazon) [inline prov/efa/src/efa_cq.c:377]: Can we put `efa_cq` to `efa_unit_test_resource`?

File: prov/efa/src/efa_cq.c [hunk 14]
```diff
@@ -331,11 +359,210 @@ int efa_cq_open(struct fid_domain *domain_fid, struct fi_cq_attr *attr,
...
+
+	assert_non_null(s);
+
+	rc = setup_readerr(s);
+	assert_int_equal(rc, 0);
+
+	s->efa_cq->ibv_cq_ex->wr_id = 0;
+	s->efa_cq->ibv_cq_ex->status = 1;
+
+	rc = efa_cq_readerr(&s->efa_cq->util_cq.cq_fid, s->entry, 0);
```
Comment (wzamazon) [inline prov/efa/src/efa_cq.c:475]: Do you need `entry` in the state? would a local variable work for this case?

**Review by wzamazon — CHANGES_REQUESTED** [review]

File: prov/efa/src/efa_cq.c (hunk 9 above)
Comment (wenduwan) [reply prov/efa/src/efa_cq.c:393]: The return code from setup and teardown hooks is not all that useful.

If we really care about the error trace I think it's better to use asserts in-place, e.g. assert not null.

File: prov/efa/test/efa_unit_test_utils.h (hunk 11 above)
Comment (wenduwan) [reply prov/efa/test/efa_unit_test_utils.h:72]: IIUC this creates a dependency on efa device. For unit test we can deep mock the data structures directly. With this deep mock we can run the test anywhere.

File: prov/efa/test/efa_unit_test_utils.h (hunk 12 above)
Comment (wenduwan) [reply prov/efa/test/efa_unit_test_utils.h:10]: Same as above.

File: prov/efa/src/efa_cq.c (hunk 13 above)
Comment (wenduwan) [reply prov/efa/src/efa_cq.c:377]: Same as above. Avoid hard dependency on efa device.

File: prov/efa/src/efa_cq.c (hunk 14 above)
Comment (wenduwan) [reply prov/efa/src/efa_cq.c:475]: We can. But I think it's encapsulated in the setup function that's shared with other test cases.

File: prov/efa/test/efa_unit_test_utils.h (hunk 12 above)
Comment (wzamazon) [reply prov/efa/test/efa_unit_test_utils.h:10]: all the other tests require devices any way, and there is no way to just run this test, so I say put it in for simplicity.

File: prov/efa/src/efa_cq.c (hunk 14 above)
Comment (wzamazon) [reply prov/efa/src/efa_cq.c:475]: I am not sure why it is better. To encapsulate it, you had to introduce a new struct.

File: prov/efa/src/rxr/rxr_ep.c [hunk 15]
```diff
@@ -2638,3 +2645,131 @@ void rxr_ep_record_tx_op_completed(struct rxr_ep *ep, struct rxr_pkt_entry *pkt_
...
+	teardown_efa_cq(state->efa_cq);
+	free(state->efa_cq);
+	free(state->efa_ep);
+	free(state->rxr_ep);
+}
+
+void __wrap_rxr_pkt_handle_send_completion(struct rxr_ep *ep,
+										   struct rxr_pkt_entry *pkt_entry)
+{
+	return mock();
```
Comment (wzamazon) [inline prov/efa/src/rxr/rxr_ep.c:2710]: Do you need to return in a "void" function?

Comment (wenduwan) [reply prov/efa/src/rxr/rxr_ep.c:2710]: Nope `mock()` is sufficient.

File: prov/efa/src/efa_cq.c (hunk 14 above)
Comment (wenduwan) [reply prov/efa/src/efa_cq.c:475]: A new struct is a good thing. We can set up required variables, etc. in a single call to `setup_xxx`, and reuse it for different test cases, e.g. write once use everywhere.

File: prov/efa/test/efa_unit_test_utils.h (hunk 12 above)
Comment (wenduwan) [reply prov/efa/test/efa_unit_test_utils.h:10]: I have a hard opinion about unit test, especially around mocking.

If possible we should make these test executable regardless of the platform, and each of our test unit should adhere to its respective test boundary. If we are testing function `void ABC(struct A, struct B, struct C)`, we should produce mocked `A, B, C` for the test, rather than introducing another dependency that is not out of the test boundary.

For example, when we test `ssize_t efa_cq_readfrom(struct fid_cq *cq_fid, void *buf, size_t count, fi_addr_t *src_addr)`, we should not introduce an external API, i.e. `fi_cq_open` which is outside our test boundary. If `fi_cq_open` is broken, that will cause `efa_cq_readfrom` tests to fail - which is against the rationale of unit testing.

File: prov/efa/src/efa_cq.c (hunk 14 above)
Comment (wzamazon) [reply prov/efa/src/efa_cq.c:463]: > A new struct is a good thing.

Not necessary.

> We can set up required variables, etc. in a single call to setup_xxx

But `entry` does not need to be setup.

File: prov/efa/test/efa_unit_test_common.c [hunk 16]
```diff
@@ -22,20 +23,81 @@ int efa_unit_test_resource_construct(struct efa_resource* resource)
 		return ret;
 	}
 
+	ret = fi_endpoint(resource->domain, resource->info, &resource->ep, NULL);
+	if (ret) {
+		fi_close(&resource->domain->fid);
```
Comment (wzamazon) [inline prov/efa/test/efa_unit_test_common.c:28]: It seems we can call `efa_unit_test_resource_destroy` on the error handling path?

**Review by wzamazon — COMMENTED** [review]
Thank you for this significant effor! I really appreciate it.

File: prov/efa/test/efa_unit_test_common.c (hunk 16 above)
Comment (wenduwan) [reply prov/efa/test/efa_unit_test_common.c:28]: Good point. Will address.

File: prov/efa/test/efa_unit_test_ep.c [hunk 17]
```diff
@@ -1,4 +1,118 @@
 #include "efa_unit_tests.h"
+#include "efa_unit_test_utils.h"
+
+static void setup_dgram_ep_resources(struct efa_resource *resource)
```
[the remark this answers is not in the document]
Comment (wenduwan) [reply prov/efa/test/efa_unit_test_ep.c:4]: Sure. Will address.

File: prov/efa/test/efa_unit_test_ep.c [hunk 18]
```diff
@@ -1,4 +1,118 @@
...
+	assert_int_equal(err, 0);
+
+	/* TX and RX are symmetrical from the progress engine POV. We only test RX. */
+	fi_ep_bind(resource->ep, &resource->cq->fid, FI_RECV);
+	fi_ep_bind(resource->ep, &resource->av->fid, 0);
+
+	efa_ep = container_of(resource->ep, struct efa_ep, util_ep.ep_fid);
+	efa_ep->info->caps &= ~FI_SEND;
+
+	efa_ep->rcq->ibv_cq_ex->start_poll = &_start_poll;
```
[the remark this answers is not in the document]
Comment (wenduwan) [reply prov/efa/test/efa_unit_test_ep.c:36]: Sure. Will address.

File: prov/efa/test/efa_unit_test_ep.c [hunk 19]
```diff
@@ -1,4 +1,118 @@
...
+
+	expect_any(__wrap_ibv_create_ah, pd);
+	expect_any(__wrap_ibv_create_ah, attr);
+	will_return(__wrap_ibv_create_ah, &ibv_ah);
+
+	err = fi_enable(resource->ep);
+	if (err) {
+		/* Don't leak reource */
+		efa_unit_test_resource_destroy(resource);
+		assert_int_equal(err, 0);
```
[the remark this answers is not in the document]
Comment (wenduwan) [reply prov/efa/test/efa_unit_test_ep.c:58]: Could you elaborate?

My intention is to destroy the resource and then raise an assertion error. This way the log will print the exact line number to help debugging.

File: prov/efa/test/efa_unit_test_ep.c [hunk 20]
```diff
@@ -1,4 +1,118 @@
...
+	efa_cq->ibv_cq_ex->read_slid = &_read_slid;
+	efa_cq->ibv_cq_ex->read_imm_data = &_read_imm_data;
+	efa_cq->ibv_cq_ex->read_opcode = &_read_opcode;
+	efa_cq->read_entry = &_read_entry;
+
+	expect_any(__wrap_ibv_create_ah, pd);
+	expect_any(__wrap_ibv_create_ah, attr);
+	will_return(__wrap_ibv_create_ah, &ibv_ah);
+
+	err = fi_enable(resource->ep);
```
[the remark this answers is not in the document]
Comment (wenduwan) [reply prov/efa/test/efa_unit_test_ep.c:109]: I think it's better to keep it separate.

`efa_unit_test_resource_construct` is a generic function. It's not tied to a particular test, and not every test needs to enable the endpoint. In order to do `fi_enable(...)` the user is required to provide additional information, i.e. configure endpoint, decide whether to mock, what to do if enable fails (some tests don't care).

File: prov/efa/test/efa_unit_test_ep.c [hunk 21]
```diff
@@ -129,14 +242,292 @@ void test_rxr_ep_dc_atomic_error_handling()
...
+	will_return_maybe(_read_opcode, IBV_WC_RECV);
+	will_return(__wrap_rxr_pkt_handle_recv_error, NULL);
+	expect_value(__wrap_rxr_pkt_handle_recv_error, ep, rxr_ep);
+	expect_value(__wrap_rxr_pkt_handle_recv_error, pkt_entry, &entry);
+	expect_any(__wrap_rxr_pkt_handle_recv_error, err);
+	expect_any(__wrap_rxr_pkt_handle_recv_error, prov_errno);
+
+	rxr_ep->util_ep.progress(&rxr_ep->util_ep);
+
+	will_return_maybe(__wrap_ibv_destroy_ah, 0);
```
[the remark this answers is not in the document]
Comment (wenduwan) [reply prov/efa/test/efa_unit_test_ep.c:531]: Because `fi_enable(...)` indirectly called `ibv_create_ah`mock, therefore we might need to mock the destruction function during resource cleanup.

File: prov/efa/test/efa_unit_test_mocks.c [hunk 22]
```diff
@@ -0,0 +1,72 @@
...
+
+void __wrap_rxr_pkt_handle_recv_completion(struct rxr_ep *ep,
+										   struct rxr_pkt_entry *pkt_entry,
+										   enum rxr_lower_ep_type lower_ep_type)
+{
+	check_expected(ep);
+	check_expected(pkt_entry);
+	check_expected(lower_ep_type);
+
+	if ((int)mock() == 42) {
```
[the remark this answers is not in the document]
Comment (wenduwan) [reply prov/efa/test/efa_unit_test_mocks.c:33]: Because of [memes](https://www.buzzfeed.com/australianredcrossbloodservice/definitive-proof-number-42-actually-is-the-answer-to-life). I have another idea in mind w.r.t. mocks in general. I put 42 as a placeholder for now. Will refactor the mocks later.

It shouldn't matter unless someone want to explicitly return `FI_ENOMSG`.

File: prov/efa/test/efa_unit_test_ep.c [hunk 23]
```diff
@@ -129,14 +242,292 @@ void test_rxr_ep_dc_atomic_error_handling()
...
+
+	/* Read 1 entry from CQ and encounter a bad wc status */
+	will_return(_start_poll, 0);
+	will_return(_end_poll, NULL);
+	efa_ep->rcq->ibv_cq_ex->status = 1;
+
+	efa_ep->util_ep.progress(&efa_ep->util_ep);
+
+	/* Error entry will be written to efa_ep->rcq->util_cq.aux_queue */
+	err = !slist_empty(&efa_ep->rcq->util_cq.aux_queue);
```
[the remark this answers is not in the document]
Comment (wenduwan) [reply prov/efa/test/efa_unit_test_ep.c:331]: As we discussed earlier, `efa_cq_readerr` is an unstable implementation detail prone to change. Therefore I chose not to call it or `efa_cq_readfrom` directly. Instead, we make this assertion to make sure the bad wc is captured.
```
...
/* Error entry will be written to efa_ep->rcq->util_cq.aux_queue */
err = !slist_empty(&efa_ep->rcq->util_cq.aux_queue);
...
assert_int_equal(err, 1);
```

If you care about `err` or `prov_errno`, we can peek into `aux_queue` instead.

File: prov/efa/test/efa_unit_test_ep.c [hunk 24]
```diff
@@ -129,14 +242,292 @@ void test_rxr_ep_dc_atomic_error_handling()
...
+
+	/* Cleanup resource before assertion */
+	will_return_maybe(__wrap_ibv_destroy_ah, 0);
+	efa_unit_test_resource_destroy(&resource);
+
+	assert_int_equal(err, 1);
+}
+
+/* Verify that RDM endpoint progress works with normal send work completions */
+void test_rdm_ep_progress_send_completion_happy(void **state)
```
[the remark this answers is not in the document]
Comment (wenduwan) [reply prov/efa/test/efa_unit_test_ep.c:341]: I'm personally ok either way.

cmocka official docs use this function signature explicitly in their user guide, e.g.
```
static void test_driver(void **state)
{
    expect_string(chef_cook, order, "hotdog");
}
```

File: prov/efa/test/efa_unit_test_ep.c (hunk 19 above)
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/test/efa_unit_test_ep.c:58]: I mean `assert_int_eqaul(err,0)` will always fail in the branch

File: prov/efa/test/efa_unit_test_mocks.c (hunk 22 above)
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/test/efa_unit_test_mocks.c:33]: Maybe use a macro like EFA_UNIT_TEST_CALL_REAL, and change the value to 4242.

> someone want to explicitly return FI_ENOMSG.

This can happen in some test case.

File: prov/efa/test/efa_unit_test_ep.c (hunk 24 above)
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/test/efa_unit_test_ep.c:341]: Unused variable will result in compiler warning. so we want to avoid it.

File: prov/efa/test/efa_unit_test_ep.c (hunk 23 above)
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/test/efa_unit_test_ep.c:331]: Yes, `efa_cq_readerr` will change.

but `fi_cq_read` is a stable API and will not change.

**Comment by shijin-aws** [comment]
could u separate your commits into the unit test and the efa provider change?

File: prov/efa/test/efa_unit_test_ep.c (hunk 19 above)
[the remark this answers is not in the document]
Comment (wenduwan) [reply prov/efa/test/efa_unit_test_ep.c:58]: Yes. That's exactly my intention. Clean up the resource and fail the test immediately.

[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/test/efa_unit_test_ep.c:58]: I see. I would move this line out of the `if` branch then.

File: prov/efa/test/rdma_core_mocks.c [hunk 25]
```diff
@@ -30,7 +17,11 @@ int __real_ibv_destroy_ah(struct ibv_ah *ibv_ah);
 
 int __wrap_ibv_destroy_ah(struct ibv_ah *ibv_ah)
 {
-	return (ibv_ah->context == NULL) ? 0 : __real_ibv_destroy_ah(ibv_ah);
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/test/rdma_core_mocks.c:33]: Still not sure why we need to get return value for this function.

File: prov/efa/test/efa_unit_test_ep.c (hunk 23 above)
[the remark this answers is not in the document]
Comment (wenduwan) [reply prov/efa/test/efa_unit_test_ep.c:331]: Makes sense. I will use it to inspect the error entry.

File: prov/efa/test/efa_unit_test_ep.c (hunk 24 above)
[the remark this answers is not in the document]
Comment (wenduwan) [reply prov/efa/test/efa_unit_test_ep.c:341]: Sure. Will address.

libfabric might have disabled this warning. Functions like `ssize_t ofi_cq_readerr(struct fid_cq *cq_fid, struct fi_cq_err_entry *buf, uint64_t flags)` have unused parameters, e.g. `flags`.

File: prov/efa/test/efa_unit_test_mocks.c (hunk 22 above)
[the remark this answers is not in the document]
Comment (wenduwan) [reply prov/efa/test/efa_unit_test_mocks.c:33]: Ok I will hard code it to 4242 for now.

File: prov/efa/test/efa_unit_test_ep.c (hunk 19 above)
[the remark this answers is not in the document]
Comment (wenduwan) [reply prov/efa/test/efa_unit_test_ep.c:58]: I'm ok to do that. It's just that we are doing 2 comparisons if `err == 0`.

**Comment by wenduwan** [comment]
> could u separate your commits into the unit test and the efa provider change?

I can but from my experience every change should be self-contained, i.e. you make the change and you prove it works. Producing separate commit makes this opaque.

Do you have a strong opinion?

**Comment by shijin-aws** [comment]
i see. i don't have a strong opinion here.

> > could u separate your commits into the unit test and the efa provider change?
> 
> I can but from my experience every change should be self-contained, i.e. you make the change and you prove it works. Producing separate commit makes this opaque.
> 
> Do you have a strong opinion?

File: prov/efa/test/rdma_core_mocks.c (hunk 25 above)
[the remark this answers is not in the document]
Comment (wenduwan) [reply prov/efa/test/rdma_core_mocks.c:33]: As it is, this mock does not really give the user control over its behavior. For example, if the user wants this function to return 123, there is no obvious way to do it.

It's much symmetrical and easier to formulate these mocks as:
```
// Inside the mock
...
val = mock();

if (val == something) {
    call_real();
} else if (val ==  something_else) {
    call_another_func();
}

...

return val;
```

File: prov/efa/test/efa_unit_test_ep.c [hunk 26]
```diff
@@ -129,14 +242,292 @@ void test_rxr_ep_dc_atomic_error_handling()
...
+	setup_dgram_ep_resources(&resource);
+
+	efa_ep = container_of(resource.ep, struct efa_ep, util_ep.ep_fid);
+
+	/* Read 1 entry from CQ and encounter a bad wc status */
+	will_return(_start_poll, 0);
+	will_return(_end_poll, NULL);
+	efa_ep->rcq->ibv_cq_ex->status = 1;
+
+	efa_ep->util_ep.progress(&efa_ep->util_ep);
```
[the remark this answers is not in the document]
Comment (wenduwan) [reply prov/efa/test/efa_unit_test_ep.c:328]: IIUC cq read is mostly used in the progress engine. So I assume testing the progress engine better simulates real-world use cases, and it covers more logic than cq read alone.

**Review by wzamazon — APPROVED** [review]
